### DISCUSSION TITLE
Variable size divisions both in X and Y at the same time

### DISCUSSION THREAD

**Discussion by beding** [kennetek/gridfinity-rebuilt-openscad#29] [Q&A]
The pictures show that this should be possible, but i do not see how.
Can somebody show me an example?
![VariableDivisionsXY_howto](https://user-images.githubusercontent.com/50421045/206904323-a89cf3c8-b844-48fc-a84b-f69b96925d35.jpg)

**Comment by kennetek** [comment]
That exact bin is the second example in `gridfinity-rebuilt-bins.scad` in the examples section at the bottom of the file. For ease of access (in case it changes in the future for some reason) I'll put the code here as well: 

```
gridfinityInit(3, 3, height(6), 0, 42) {
    cut(x=0, y=0, w=1.5, h=0.5, t=5, s=false);
    cut(0, 0.5, 1.5, 0.5, 5, false);
    cut(0, 1, 1.5, 0.5, 5, false);
    
    cut(0,1.5,0.5,1.5,5,false);
    cut(0.5,1.5,0.5,1.5,5,false);
    cut(1,1.5,0.5,1.5,5,false);
    
    cut(1.5, 0, 1.5, 5/3, 2);
    cut(1.5, 5/3, 1.5, 4/3, 4);
}
gridfinityBase(3, 3, 42, 0, 0, 1);
```

Documentation of the functions used here can also be found at the project [website](https://kennetek.github.io/gridfinity-rebuilt-openscad/).
Hope that helps!

**Comment by beding** [comment]
Hi Kenneth,

Thanks for your response.

I am an experienced C++ programmer, but I know nothing about OpenSCAD.

So, I rely on the parameters that I can set in the UI.

Can you tell me How can I use this function?

Regards,

Bert

 

Van: Kenneth ***@***.***> 
Verzonden: zondag 11 december 2022 23:58
Aan: kennetek/gridfinity-rebuilt-openscad ***@***.***>
CC: beding ***@***.***>; Author ***@***.***>
Onderwerp: Re: [kennetek/gridfinity-rebuilt-openscad] Variable size divisions both in X and Y at the same time (Discussion #29)

 

That exact bin is the second example in gridfinity-rebuilt-bins.scad in the examples section at the bottom of the file. For ease of access (in case it changes in the future for some reason) I'll put the code here as well:

gridfinityInit(3, 3, height(6), 0, 42) {
    cut(x=0, y=0, w=1.5, h=0.5, t=5, s=false);
    cut(0, 0.5, 1.5, 0.5, 5, false);
    cut(0, 1, 1.5, 0.5, 5, false);
    
    cut(0,1.5,0.5,1.5,5,false);
    cut(0.5,1.5,0.5,1.5,5,false);
    cut(1,1.5,0.5,1.5,5,false);
    
    cut(1.5, 0, 1.5, 5/3, 2);
    cut(1.5, 5/3, 1.5, 4/3, 4);
}
gridfinityBase(3, 3, 42, 0, 0, 1);

Documentation of the functions used here can also be found at the project website <https://kennetek.github.io/gridfinity-rebuilt-openscad/> .
Hope that helps!

**Reply by Ruudjhuu** [reply]
Hi @beding . I also am more experienced with c and c++. The main difference of the openscad language is that it is a functional language and not OOP as c++.

You can see the functions called inside the scope of a function (curly brackets) as additional function pointer arguments (callbacks). These functions (`cut` in this case) are called inside the `GridfinityInit` scope. These callbacks can be used by the "parent" function using the `children()` function. 

So the cut functions wil create blocks on specific locations, and because these blocks are children of `GridfinityInit` the blocks will be used with the `difference` function which will remove them from the initial gridfinity block. (See the code in the `GridfinityInit` function for details)

In OpenScad you can use "!" and "#" to debug. "!" will only show the result of the line you put it in front and "#" will highlight the part that the line is responsible for if possible. So if you put a "!" Infront of one of the `cut` functions, you`ll see what is happening.
